### ISSUE TITLE
RestoreTask is not "thread safe" - NuGet restore fails with "Cannot create a file when that file already exists" when restore invoked in parallel

### ISSUE THREAD

**Issue by natemcmaster** [NuGet/Home#7648]
Invoking the "Restore" target on projects in parallel leads to a race condition in committing the restore result to disk. This appears most often as the result of developers writing their own MSBuild targets and using the `<MSBuild>` task to invoke Restore.

The only workaround I could find was to avoid this calling pattern completely. Instead, you can call the the 'restore' target in NuGet.targets and pass in the list of projects.

cref https://github.com/dotnet/arcade/pull/1567
cref https://github.com/aspnet/BuildTools/blob/0872134136bf3d181e3ca3fa700b7ff392c055b3/files/KoreBuild/modules/projectbuild/module.targets#L21-L70

## Ideas
I don't see an obvious solution. Here are some ideas

* Use `IBuildEngine4.GetRegisteredTaskObject` within `RestoreTask` to avoid double-restores during the same build. 
* Use `IBuildEngine4.GetRegisteredTaskObject` within `RestoreTask` to detect a double-restore was invoked, and provide a warning message.
* Include _which_ file is the problem in the error message.
* Instead of failing immediately when you can't write the restore result, retry.

## Details about Problem
NuGet product used: dotnet.exe or MSBuild.exe
dotnet.exe --version: 2.2.100
VS version: 15.9
OS version: Any
Worked before?: no

## Detailed repro steps so we can see the same problem

1. Create a set of 3 simple .NET standard class library projects (see attached repro). Two projects should both have ProjectReference to the third project.
```
Lib A --> LibC
Lib B --> LibC
```
2. Create an MSBuild targets file like this. We'll call it "ci.proj".
```xml
<!-- ci.proj -->
<Project>
    <ItemGroup>
        <Projects Include="LibA/LibA.csproj" />
        <Projects Include="LibB/LibB.csproj" />
        <Projects Include="LibC/LibC.csproj" />
    </ItemGroup>

    <Target Name="Restore">
        <MSbuild Projects="@(Projects)"
                 Targets="Restore"
                 BuildInParallel="true" />
    </Target>
</Project>
```
3. Run restore by invoking `msbuild -t:ci.proj -t:restore`

## Result

NuGet fails in LibC with an error that says "Cannot create a file when that file already exists". The error message does not indicate which file was the problem.

```
 Restoring packages for C:\tmp\restoreRepro\LibC\LibC.csproj...
  Restoring packages for C:\tmp\restoreRepro\LibC\LibC.csproj...
  Restoring packages for C:\tmp\restoreRepro\LibB\LibB.csproj...
  Generating MSBuild file C:\tmp\restoreRepro\LibC\obj\LibC.csproj.nuget.g.props.
  Generating MSBuild file C:\tmp\restoreRepro\LibC\obj\LibC.csproj.nuget.g.targets.
  Generating MSBuild file C:\tmp\restoreRepro\LibC\obj\LibC.csproj.nuget.g.targets.
  Generating MSBuild file C:\tmp\restoreRepro\LibB\obj\LibB.csproj.nuget.g.props.
  Generating MSBuild file C:\tmp\restoreRepro\LibB\obj\LibB.csproj.nuget.g.targets.
  Restore completed in 12.8 ms for C:\tmp\restoreRepro\LibC\LibC.csproj.
  Restore completed in 14.61 ms for C:\tmp\restoreRepro\LibB\LibB.csproj.
  Restoring packages for C:\tmp\restoreRepro\LibA\LibA.csproj...
  Restoring packages for C:\tmp\restoreRepro\LibC\LibC.csproj...
C:\Users\namc\.dotnet\x64\sdk\2.2.100\NuGet.targets(114,5): error : Cannot create a file when that file already exists [C:\tmp\restoreRepro\LibB\LibB.csproj]
  Generating MSBuild file C:\tmp\restoreRepro\LibA\obj\LibA.csproj.nuget.g.props.
  Generating MSBuild file C:\tmp\restoreRepro\LibA\obj\LibA.csproj.nuget.g.targets.
  Restore completed in 128.42 ms for C:\tmp\restoreRepro\LibC\LibC.csproj.
  Restore completed in 147.55 ms for C:\tmp\restoreRepro\LibA\LibA.csproj.
```

## Other suggested things

This is a race condition, so it may not appear right away. But it is consistent. I get consistent repros like this:
```ps1
for($i = 0; $i -lt 100; $i+=1) {
  gci -re */obj/**/* | % {rm $_ } # clean the last restore result
  dotnet msbuild build.proj -bl -t:Restore
  if ($LASTEXITCODE -ne 0){ break }
}
```
### Verbose Logs

[msbuild.binlog.zip](https://github.com/NuGet/Home/files/2688507/msbuild.binlog.zip)

### Sample Project

[repro-parallel-restore.zip](https://github.com/NuGet/Home/files/2688501/repro-parallel-restore.zip)

**Comment by rainersigwald** [comment]
I think there's another problem here beyond the race: if I understand the restore operation correctly, the _results of restore_ can differ when invoked for different project graph subsets, because a supergraph can result in a different solution to the diamond-dependency-resolved closure of package references.

**Comment by nkolev92** [comment]
@rainersigwald 
I'm not sure I fully follow, but I don't think that's the case. 
The restore operation on each project requires the full closure of projects to be "discoverable", regardless whether those transitive projects are getting restored themselves or not.

**Comment by rocifier** [comment]
Nice report and repro. I’m curious if this is just an issue with .net core projects or if the older package format also has this problem.

**Comment by ViktorHofer** [comment]
This is fixed with the new RestoreTaskEx path which can be enabled via the `/p:RestoreUseStaticGraphEvaluation=true` switch.

**Comment by nkolev92** [comment]
@ViktorHofer 
Seems unlikely.
Static graph restore changes the msbuild part, the rest of it is mostly the same. 

I'd need to investigate to confirm though.

**Comment by ViktorHofer** [comment]
You are right, I just hit this again with RestoreTaskEx :(

**Comment by ViktorHofer** [comment]
Here's the call stack:

```
Cannot create a file when that file already exists.
   at System.IO.FileSystem.MoveFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at System.IO.File.Move(String sourceFileName, String destFileName)
   at NuGet.Common.FileUtility.Move(String sourceFileName, String destFileName)
   at NuGet.Common.FileUtility.Replace(String sourceFileName, String destFileName)
   at NuGet.Common.FileUtility.Replace(Action`1 writeSourceFile, String destFilePath)
   at NuGet.Commands.RestoreResult.CommitAssetsFileAsync(LockFileFormat lockFileFormat, IRestoreResult result, ILogger log, Boolean toolCommit, CancellationToken token)
   at NuGet.Commands.RestoreResult.CommitAsync(ILogger log, CancellationToken token)
   at NuGet.Commands.RestoreRunner.CommitAsync(RestoreResultPair restoreResult, CancellationToken token)
   at NuGet.Commands.RestoreRunner.ExecuteAndCommitAsync(RestoreSummaryRequest summaryRequest, CancellationToken token)
   at NuGet.Commands.RestoreRunner.CompleteTaskAsync(List`1 restoreTasks)
   at NuGet.Commands.RestoreRunner.RunAsync(IEnumerable`1 restoreRequests, RestoreArgs restoreContext, CancellationToken token)
   at NuGet.Commands.RestoreRunner.RunAsync(RestoreArgs restoreContext, CancellationToken token)
   at NuGet.Build.Tasks.BuildTasksUtility.RestoreAsync(DependencyGraphSpec dependencyGraphSpec, Boolean interactive, Boolean recursive, Boolean noCache, Boolean ignoreFailedSources, Boolean disableParallel, Boolean force, Boolean forceEvaluate, Boolean hideWarningsAndErrors, Boolean restorePC, ILogger log, CancellationToken cancellationToken)
   at NuGet.Build.Tasks.Console.MSBuildStaticGraphRestore.RestoreAsync(String entryProjectFilePath, IDictionary`2 globalProperties, IReadOnlyDictionary`2 options)
```

**Comment by nkolev92** [comment]
Thanks @ViktorHofer That's super helpful.
We have a mechanism for locking on that but it could often be expensive. 

We've focused on parallelization of package install operations cause those are a lot more infrequent. 
We haven't used the expensive locking mechanism for the project specific assets.

**Comment by ViktorHofer** [comment]
I think you should be able to force this to happen locally with the repro that @natemcmaster posted. @natemcmaster implemented a [workaround for that a year ago in Arcade](https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj#L214-L246) but with the new RestoreTaskEx path, this doesn't work anymore which basically means we are blocked turning on the new static graph restore until this is fixed.

Let me know if I can help you debug this somehow. I also just pinged @jeffkl regarding this offline.

**Comment by ViktorHofer** [comment]
Ok @jeffkl pointed out that this could be related to us having multiple files with the same name, ie `packages.proj` which are now binclashing in the intermediate dir.

**Comment by jeffkl** [comment]
Yeah I'm pretty confident after looking at the code (which does a delete then move) that two or more projects are sharing the same restore output path which is not supported.  I know Arcade defaults to using a rooted BaseIntermediateOutput path so you have to ensure that every project in the tree has a unique file name.

NuGet could add an error and fail restore if two projects collide as well

**Comment by ViktorHofer** [comment]
Happened again for a uniquely named project:

```
  Restored F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj (in 6.04 sec).
  Restored F:\workspace\_work\1\s\src\libraries\System.Reflection.Extensions\tests\System.Reflection.Extensions.Tests.csproj (in 101 ms).
  Restored F:\workspace\_work\1\s\src\libraries\System.Reflection.Emit\tests\System.Reflection.Emit.Tests.csproj (in 67 ms).
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error : Cannot create a file when that file already exists. [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at System.IO.FileSystem.MoveFile(String sourceFullPath, String destFullPath, Boolean overwrite) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at NuGet.Common.FileUtility.Move(String sourceFileName, String destFileName) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at NuGet.Common.FileUtility.Replace(String sourceFileName, String destFileName) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at NuGet.Common.FileUtility.Replace(Action`1 writeSourceFile, String destFilePath) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.BuildAssetsUtils.WriteXML(String path, XDocument doc) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.BuildAssetsUtils.WriteFiles(IEnumerable`1 files, ILogger log) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.RestoreResult.CommitAssetsFileAsync(LockFileFormat lockFileFormat, IRestoreResult result, ILogger log, Boolean toolCommit, CancellationToken token) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.RestoreResult.CommitAsync(ILogger log, CancellationToken token) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.RestoreRunner.CommitAsync(RestoreResultPair restoreResult, CancellationToken token) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.RestoreRunner.ExecuteAndCommitAsync(RestoreSummaryRequest summaryRequest, CancellationToken token) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.RestoreRunner.CompleteTaskAsync(List`1 restoreTasks) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.RestoreRunner.RunAsync(IEnumerable`1 restoreRequests, RestoreArgs restoreContext, CancellationToken token) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at NuGet.Commands.RestoreRunner.RunAsync(RestoreArgs restoreContext, CancellationToken token) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at NuGet.Build.Tasks.BuildTasksUtility.RestoreAsync(DependencyGraphSpec dependencyGraphSpec, Boolean interactive, Boolean recursive, Boolean noCache, Boolean ignoreFailedSources, Boolean disableParallel, Boolean force, Boolean forceEvaluate, Boolean hideWarningsAndErrors, Boolean restorePC, ILogger log, CancellationToken cancellationToken) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error :    at NuGet.Build.Tasks.Console.MSBuildStaticGraphRestore.RestoreAsync(String entryProjectFilePath, IDictionary`2 globalProperties, IReadOnlyDictionary`2 options) [F:\workspace\_work\1\s\src\installer\test\Microsoft.NET.HostModel.Tests\Microsoft.NET.HostModel.AppHost.Tests\Microsoft.NET.HostModel.AppHost.Tests.csproj]
##[error].dotnet\sdk\5.0.100-preview.3.20170.6\NuGet.RestoreEx.targets(10,5): error : (NETCORE_ENGINEERING_TELEMETRY=Restore) Cannot create a file when that file already exists.
   at System.IO.FileSystem.MoveFile(String sourceFullPath, String destFullPath, Boolean overwrite)
   at NuGet.Common.FileUtility.Move(String sourceFileName, String destFileName)
   at NuGet.Common.FileUtility.Replace(String sourceFileName, String destFileName)
   at NuGet.Common.FileUtility.Replace(Action`1 writeSourceFile, String destFilePath)
   at NuGet.Commands.BuildAssetsUtils.WriteXML(String path, XDocument doc)
   at NuGet.Commands.BuildAssetsUtils.WriteFiles(IEnumerable`1 files, ILogger log)
   at NuGet.Commands.RestoreResult.CommitAssetsFileAsync(LockFileFormat lockFileFormat, IRestoreResult result, ILogger log, Boolean toolCommit, CancellationToken token)
   at NuGet.Commands.RestoreResult.CommitAsync(ILogger log, CancellationToken token)
   at NuGet.Commands.RestoreRunner.CommitAsync(RestoreResultPair restoreResult, CancellationToken token)
   at NuGet.Commands.RestoreRunner.ExecuteAndCommitAsync(RestoreSummaryRequest summaryRequest, CancellationToken token)
   at NuGet.Commands.RestoreRunner.CompleteTaskAsync(List`1 restoreTasks)
   at NuGet.Commands.RestoreRunner.RunAsync(IEnumerable`1 restoreRequests, RestoreArgs restoreContext, CancellationToken token)
   at NuGet.Commands.RestoreRunner.RunAsync(RestoreArgs restoreContext, CancellationToken token)
   at NuGet.Build.Tasks.BuildTasksUtility.RestoreAsync(DependencyGraphSpec dependencyGraphSpec, Boolean interactive, Boolean recursive, Boolean noCache, Boolean ignoreFailedSources, Boolean disableParallel, Boolean force, Boolean forceEvaluate, Boolean hideWarningsAndErrors, Boolean restorePC, ILogger log, CancellationToken cancellationToken)
   at NuGet.Build.Tasks.Console.MSBuildStaticGraphRestore.RestoreAsync(String entryProjectFilePath, IDictionary`2 globalProperties, IReadOnlyDictionary`2 options)
```

https://dev.azure.com/dnceng/public/_build/results?buildId=575356&view=logs

**Comment by ViktorHofer** [comment]
Definitely looks like a race as you can see that the same project was restored above.

**Comment by jeffkl** [comment]
It kind of looks like multiple restores are happening at the same time for the same project?  Is that true?

**Comment by ViktorHofer** [comment]
Those are transitive dependencies, ie [AppHost.Bundle.Tests.csproj](https://github.com/dotnet/runtime/blob/7d67d17a9f49ad5f365467fcd3bf0d25f2b9349a/src/installer/test/Microsoft.NET.HostModel.Tests/AppHost.Bundle.Tests/AppHost.Bundle.Tests.csproj#L16) references `Microsoft.NET.HostModel.csproj` which then gets restored multiple times:

```
  Determining projects to restore...
  Determining projects to restore...
  Determining projects to restore...
  Determining projects to restore...
  Determining projects to restore...
  Determining projects to restore...
  Restored C:\git\runtime4\src\installer\managed\Microsoft.NET.HostModel\Microsoft.NET.HostModel.csproj (in 672 ms).
  Restored C:\git\runtime4\src\installer\managed\Microsoft.NET.HostModel\Microsoft.NET.HostModel.csproj (in 906 ms).
  Restored C:\git\runtime4\src\installer\managed\Microsoft.NET.HostModel\Microsoft.NET.HostModel.csproj (in 1.15 sec).
  Restored C:\git\runtime4\src\installer\managed\Microsoft.NET.HostModel\Microsoft.NET.HostModel.csproj (in 937 ms).
  Restored C:\git\runtime4\src\installer\managed\Microsoft.NET.HostModel\Microsoft.NET.HostModel.csproj (in 1.23 sec).
```

So restoring entrypoints with same dependencies in parallel leads to the issue that we are seeing here. The obvious solution here is to move to a single restore container instead of restoring in parallel. I'll see what I can do about that in dotnet/runtime.

**Comment by jeffkl** [comment]
@nkolev92 there's parts of NuGet that support multiple restores at the same time, but I don't think we support restoring the same project at the same time right?

@ViktorHofer one massive restore for the whole repo is sort of how retail NuGet is designed.  My static graph evaluation improvements are meant to speed up the MSBuild side of things by loading projects faster.  I made other improvements to the NuGet side to speed up no-op restores as well.  I'm not sure if that's what was the goal of restoring portions of the tree in parallel, but hopefully one massive restore will be fast and not have any race conditions 👍

**Comment by nkolev92** [comment]
> there's parts of NuGet that support multiple restores at the same time, but I don't think we support restoring the same project at the same time right

Not really. The shared parts are synchronized, global packages folder, http cache, plugin cache, temp cache, but we never synchronize the project intermediates. 
Process wide synchronization is expensive so we avoid it unless absolutely necessary.

**Comment by johnpankowicz** [comment]
This problem also occurs when running multiple build tasks in Visual Studio Code. I define a "Build All" task in tasks.json with a "dependsOn" property that contains multiple other tasks. Two of these tasks load the same NuGet package. If that package is not currently in the cache, one of the dependent tasks will fail with this error.

**Comment by nkolev92** [comment]
> Two of these tasks load the same NuGet package. If that package is not currently in the cache, one of the dependent tasks will fail with this error.

Hey @johnpankowicz if you are referring to usage of packages from the global packages folder, then it's not something that NuGet monitors. 
NuGet guards the installation in the global packages folder with process wide locks.

> I define a "Build All" task in tasks.json with a "dependsOn" property that contains multiple other tasks.

Are these tasks running multiple restores?

Is dotnet restore getting called explicitly?

**Comment by johnpankowicz** [comment]
@nkolev92 - I am referring to the global packages folder. I deleted the contents to test what happens when someone clones my Github solution and has none of my NuGet packages installed.
Each of the dependent tasks just do a "dotnet build" of a project that has multiple NuGet package references. "dotnet restore" is not being run explicitly. Many of the projects reference the same Nuget packages. I need to run the "Build All" task usually twice to get around this error.

**Comment by nkolev92** [comment]
@johnpankowicz 
I'm not familiar with the details of how VS Code builds, but whenever you call `dotnet build` restore is run automatically and will detect packages missing from the global packages folder if necessary. 

If you can repro any dotnet build or dotnet restore not getting the correct packages, we'd be happy to take a look at that (best done in a separate issue).

**Comment by japj** [comment]
is this also related to https://github.com/NuGet/Home/issues/7060 ?

i.e. if there multiple build agents running restore on packages that are not yet present in the local nuget cache you might run into this issue?

Since I believe the azuredevops build agent will have its own TEMP folder, thus locking on TEMP folder might not be correct for the nuget (global) user cache folder.

**Comment by karthikrathinavelu** [comment]
Faced similar issue in our service as well albeit sporadically. In our buildtracker machines it fails sometimes with the following error

2021/05/27 10:32:07 AM        79>Restore:
2021/05/27 10:32:07 AM             Restoring packages for X:\bt\1211620\repo\src\Service\MonitoringUtilities\MonitoringUtilities.csproj...
2021/05/27 10:32:07 AM             Committing restore...
2021/05/27 10:32:07 AM             Restoring packages for X:\bt\1211620\repo\src\Service\MonitoringUtilities\MonitoringUtilities.csproj...
2021/05/27 10:32:07 AM             Assets file has not changed. Skipping assets file writing. Path: X:\bt\1211620\repo\src\Samples\HelloWebApi\.pkgrefgen\project.assets.json
2021/05/27 10:32:07 AM             Restored X:\bt\1211620\repo\src\Samples\HelloWebApi\HelloWebApi.csproj (in 162 ms).
2021/05/27 10:32:07 AM             Restoring packages for X:\bt\1211620\repo\src\Service\MonitoringControllers\MonitoringControllers.csproj...
2021/05/27 10:32:07 AM             
2021/05/27 10:32:07 AM             NuGet Config files used:
2021/05/27 10:32:07 AM                 X:\bt\1211620\repo\NuGet.Config
2021/05/27 10:32:07 AM                 C:\Users\azbldrun\AppData\Roaming\NuGet\NuGet.Config
2021/05/27 10:32:07 AM             
2021/05/27 10:32:07 AM             All projects are up-to-date for restore.
2021/05/27 10:32:07 AM             Restoring packages for X:\bt\1211620\repo\src\Service\MonitoringControllers\MonitoringControllers.csproj...
2021/05/27 10:32:07 AM             Restoring packages for X:\bt\1211620\repo\src\Service\MonitoringUtilities\MonitoringUtilities.csproj...
2021/05/27 10:32:07 AM             Restoring packages for X:\bt\1211620\repo\src\Service\Hosting\ServiceFabric\MonitoringWebService\MonitoringWebService.csproj...
2021/05/27 10:32:08 AM             Committing restore...
2021/05/27 10:32:08 AM             Committing restore...
2021/05/27 10:32:08 AM             Committing restore...
2021/05/27 10:32:08 AM             Committing restore...
2021/05/27 10:32:08 AM             Committing restore...
2021/05/27 10:32:08 AM             Writing assets file to disk. Path: X:\bt\1211620\repo\src\Service\MonitoringUtilities\.pkgrefgen\project.assets.json
2021/05/27 10:32:09 AM             Writing assets file to disk. Path: X:\bt\1211620\repo\src\Service\MonitoringControllers\.pkgrefgen\project.assets.json
2021/05/27 10:32:09 AM             Writing assets file to disk. Path: X:\bt\1211620\repo\src\Service\MonitoringUtilities\.pkgrefgen\project.assets.json
2021/05/27 10:32:09 AM             Writing assets file to disk. Path: X:\bt\1211620\repo\src\Service\MonitoringControllers\.pkgrefgen\project.assets.json
2021/05/27 10:32:09 AM             Writing assets file to disk. Path: X:\bt\1211620\repo\src\Service\MonitoringUtilities\.pkgrefgen\project.assets.json
2021/05/27 10:32:09 AM             Restored X:\bt\1211620\repo\src\Service\MonitoringUtilities\MonitoringUtilities.csproj (in 2.07 sec).
2021/05/27 10:32:09 AM             Restored X:\bt\1211620\repo\src\Service\MonitoringControllers\MonitoringControllers.csproj (in 2.07 sec).
2021/05/27 10:32:09 AM             Restored X:\bt\1211620\repo\src\Service\MonitoringUtilities\MonitoringUtilities.csproj (in 2.07 sec).
2021/05/27 10:32:09 AM             
2021/05/27 10:32:09 AM             NuGet Config files used:
2021/05/27 10:32:09 AM                 X:\bt\1211620\repo\NuGet.Config
2021/05/27 10:32:09 AM                 C:\Users\azbldrun\AppData\Roaming\NuGet\NuGet.Config
2021/05/27 10:32:09 AM             
2021/05/27 10:32:09 AM             Feeds used:
2021/05/27 10:32:09 AM                 https://msazure.pkgs.visualstudio.com/One/_packaging/AzureBackup/nuget/v3/index.json
2021/05/27 10:32:09 AM        78>X:\.BtCxCache\.A\MsBuild.Corext.5JZj-zRYrIEfXT5Ktun1vg\Current\bin\NuGet.RestoreEx.targets(19,5): error : Cannot create a file when that file already exists. [X:\bt\1211620\repo\src\Service\MonitoringControllers\MonitoringControllers.csproj]

**Comment by kpete2017** [comment]
Currently, we are still dealing with the issue in Visual Studio during its Nuget restore step. When we run a build it will come up with the duplicate file error intermittently which we can only get rid of by rebuilding the affected projects one by one. These are UWP projects with ~80 projects loaded in the solution but it seems less likely to run into this problem when there are fewer projects to restore.

**Comment by nkolev92** [comment]
> Currently, we are still dealing with the issue in Visual Studio during its Nuget restore step. When we run a build it will come up with the duplicate file error intermittently which we can only get rid of by rebuilding the affected projects one by one. These are UWP projects with ~80 projects loaded in the solution but it seems less likely to run into this problem when there are fewer projects to restore.

In Visual Studio, the RestoreTask is not run. NuGet is executed within the VS process.

I wouldn't expect you to be running in issues like the ones being discussed here. 
Do you have any customizations in how you invoke restore?

**Comment by kpete2017** [comment]
> In Visual Studio, the RestoreTask is not run. NuGet is executed within the VS process.
> 
> I wouldn't expect you to be running in issues like the ones being discussed here. Do you have any customizations in how you invoke restore?

There are no customizations to the restore whatsoever from our end. We are running React Native for Windows with its experimental NuGet feature which may possibly change how the solution is restored but I'm not sure. I can move this to an issue over on their GitHub page if you think it's more likely a problem from their end. This is just the only thread I've seen with the same "Cannot create a file when that file already exists" error coming from the restore step.

**Comment by nkolev92** [comment]
I see. Based on your comments, seems like the the issues might not be equivalent. 

> React Native for Windows with its experimental NuGet feature

I'm not really sure what does that, so can't really recommend either way.

**Comment by CDAlexanderW** [comment]
Could an additional command line option to globally lock on the path/project help? So that restoring the same project does at least not break by accessing the same files at the same time. And this wouldn't hurt anyone not needing that (performance wise) and help anyone having this exact problem.

We got the same issue with the following scenario: we run multiple nuget.exe on different sln solution files that contain the same projects, so they sometimes collide when nuget has to actually restore packages and does not really have problems when nothing really has to be done (up-to-date). It dramatically improves the performance to run nuget.exe in parallel centralized beforehand on all solutions in a repository. Not directly related with this issue: In our case having the option to add multiple sln solution files to one nuget.exe call, where nuget would mange the individual projects of them accross in an efficient way (parallel), would be even better, but I don't think that this option exists, nor that someone will implement this.

**Comment by ViktorHofer** [comment]
> In our case having the option to add multiple sln solution files to one nuget.exe call, where nuget would mange the individual projects of them accross in an efficient way (parallel), would be even better, but I don't think that this option exists, nor that someone will implement this.

What if you just have one central solution file that contains everything and then solution filter files for the projects that are currently grouped? NuGet would then only need to be invoked once.

**Comment by CDAlexanderW** [comment]
> > In our case having the option to add multiple sln solution files to one nuget.exe call, where nuget would mange the individual projects of them accross in an efficient way (parallel), would be even better, but I don't think that this option exists, nor that someone will implement this.
> 
> What if you just have one central solution file that contains everything and then solution filter files for the projects that are currently grouped? NuGet would then only need to be invoked once.

Thx for your reply and suggestion @ViktorHofer!

There are two problems with that. I am sure in the large group of developers someone would miss to change it also in the large solution file (some sln files are also located in different sub directories). And for CI builds we only build some of the sln files (you can combine them) -> we would have to have multiple of these sln files which would be another maintenance mess.

**Comment by ViktorHofer** [comment]
> I am sure in the large group of developers someone would miss to change it also in the large solution file (some sln files are also located in different sub directories).

AFAIK it's not possible to only list a project in the solution filter file as behind the scenes it really just points to the solution file and specifies which projects to include from it. The IDE automatically adds it to both and if you would manually only update the slnf, then you would hopefully get an error by tooling / nuget / etc that the slnf is invalid.

> And for CI builds we only build some of the sln files (you can combine them) -> we would have to have multiple of these sln files which would be another maintenance mess.

I understand that you have special requirements here but just throwing out another idea. With a single sln file at the root, you could restore that once and then only build the individual slnfs that you want, by passing the `--no-restore` flag in.

**Comment by CDAlexanderW** [comment]
> > I am sure in the large group of developers someone would miss to change it also in the large solution file (some sln files are also located in different sub directories).
> 
> AFAIK it's not possible to only list a project in the solution filter file as behind the scenes it really just points to the solution file and specifies which projects to include from it. The IDE automatically adds it to both and if you would manually only update the slnf, then you would hopefully get an error by tooling / nuget / etc that the slnf is invalid.
> 
I totally missed the "filter" word in your initial reply. Now I understand what you meant (I thought about using a normal sln file which is different). Yes that would work, but I am scared that I am not able to change our environment to look like that.
> > And for CI builds we only build some of the sln files (you can combine them) -> we would have to have multiple of these sln files which would be another maintenance mess.
> 
> I understand that you have special requirements here but just throwing out another idea. With a single sln file at the root, you could restore that once and then only build the individual slnfs that you want, by passing the `--no-restore` flag in.

Thx for another idea. That would work if we would use your suggestion with the root sln, but still it would restore projects not needed for that work load.
For us it still seems to be the better option at the moment to either use the half broken parallel execution we use now or simply change back to not parallel and have much longer execution times, especially when everything is up to date.

**Comment by ViktorHofer** [comment]
One more suggestion from my side. We, the .NET product often use solution files or solution files to orchestrate the product construction. In some more complex repositories like https://github.com/dotnet/runtime, we use the `Microsoft.Build.Traversal` msbuild sdk instead: https://github.com/microsoft/MSBuildSdks/tree/main/src/Traversal

This allows us to just build the parts of the product that we need with the ability to condition based on input properties. Here's an example: https://github.com/dotnet/maintenance-packages/blob/main/Build.proj

I could imagine that if you would replace all current solution files with Traversal msbuild sdk projects and then add one more Traversal file at the repo root that references all the other traversal projects, you would achieve what you want. This gives you the flexibility to only restore and build what you need based on inputs with the best performance possible.

The downside of that approach is that you can't "double-click" such msbuild projects and open them with VS unless you install an extension or use [`slngen`](https://github.com/microsoft/slngen?tab=readme-ov-file#slngen) which generates a solution file on the fly and automatically opens VS and loads it.

I'm just throwing out these ideas as I don't think that running NuGet in parallel is the right solution.
